Use timecop in safe mode

felix-dumit opened the pull request:
I enabled timecop in safe mode (https://github.com/travisjeffery/timecop#timecopsafe_mode) for a project and this gem caused failures since it's not using the block syntax for timecop.

This change makes it pass the `safe_mode` check.

felix-dumit committed:
Use timecop in safe mode

kevin-j-m commented:
Thanks for the contribution! I have a few requests & suggestions here:

1. Can you rebase against the latest `main`, particularly so this pulls the changes from #40 in? That should get us a running CI back operating on this repo.
2. I think there is another safe mode violation in `Manager#increase_time`. Do you agree & can you resolve that as well?
3. I'm fine with adding `Timecop.safe_mode = true` to the `spec/spec_helper.rb` file so that the test suite confirms this continues to be safe mode compliant. Would you mind adding that so we can demonstrate with the test suite that we're compliant with safe mode?

felix-dumit commented:
Thanks for the feedback!

1. Done, not sure how the PR picked up the wrong branch to start with, I did it through the github UI, but it should be rebased now
2. Yes that's true, I wasn't hitting that in my project so didn't notice, I fixed it as well
3. Added, specs are all passing now with safe mode enabled

kevin-j-m reviewed (approved):
Thanks again.

kevin-j-m commented:
I released version [0.5.0](https://github.com/kevin-j-m/clockwork-test/blob/main/CHANGELOG.md#050-2023-04-20) with this change. Thank you for your contribution!

felix-dumit commented:
Awesome, thanks for the quick turnaround!

robinbrandt commented on lib/clockwork/test/manager.rb:
```diff
@@ -57,21 +57,17 @@ def set_opts(opts)
-
-          @total_ticks += 1
-          break if ticks_exceeded? || time_exceeded?
+        loop.with_index do |_, index|
```
This does change the behaviour a bit; not sure if this was intentional (the PR description does not mention this).

Before this change this would work:

clock.rb:
```ruby
  # ...
  every(1.hour, "stand up") do
    puts "Stand up!"
  end
```

clock_test.rb:
```ruby
  # ...
  test "stand up" do
    time = Time.utc(2023, 4, 21)

    Clockwork::Test.run(max_ticks: 1, start_time: time, end_time: time)
    assert_equal(1, Clockwork::Test.times_run("stand up"))
  end
```

Before we'd do at least one tick if start and end date are the same. Now `#time_exceeded?` is checked at the beginning of the loop and we exit early.

kevin-j-m replied:
Thanks for bringing this up. I did not intend any functional change as as result of this.

Sounds like we're missing a test case!

@robinbrandt if you're able to submit a PR to modify this, I'm happy to review. Just a heads up that I'm heading on vacation starting today, so I'll likely be delayed in reviewing.

I appreciate you commenting here!

robinbrandt replied:
> if you're able to submit a PR to modify this

Happy to do so!

https://github.com/kevin-j-m/clockwork-test/pull/41.

> I'm heading on vacation

Enjoy your vacation!
